### PR TITLE
Test maker actor system shutdown

### PR DESCRIPTION
**Pull request by klochowicz**
- maker actor system now stops all its async tasks when it gets destroyed
- taker notices absence of maker in a newly written test

Note: I did not include the roll-out on the taker side (it would be done the
same way), as I wanted to first get some feedback from the reviewers.

### CODE REVIEW

File: daemon/src/tokio_ext.rs [hunk 1]
```diff
@@ -27,3 +28,20 @@ where
         timeout(duration, self)
     }
 }
+
+pub trait SpawnWithHandleExt: Future {
+    /// Spawn the future on a task in the runtime and return a RemoteHandle to it.
+    /// The task will be stopped when the handle gets dropped.
+    fn spawn_with_handle(self) -> RemoteHandle<Self::Output>;
```
Comment (thomaseizinger) [inline daemon/src/tokio_ext.rs]: You can put this on the `FutureExt` trait on line 19.

Creating a dedicated extension trait for each function will only bloat our imports.

File: daemon/src/lib.rs [hunk 2]
```diff
@@ -223,13 +247,15 @@ where
             n_payouts,
         )
         .create(None)
-        .spawn_global();
+        .run();
```
Comment (thomaseizinger) [inline daemon/src/lib.rs:256]: This change seems unrelated?

File: daemon/src/lib.rs [hunk 3]
```diff
@@ -190,6 +213,7 @@ where
         oracle_constructor: impl FnOnce(Vec<Cfd>, Box<dyn StrongMessageChannel<Attestation>>) -> O,
         monitor_constructor: impl FnOnce(Box<dyn StrongMessageChannel<monitor::Event>>, Vec<Cfd>) -> F,
         n_payouts: usize,
+        maker_heartbeat_interval: Duration,
```
Comment (thomaseizinger) [inline daemon/src/lib.rs:216]: Why do we need to make this configurable?

File: daemon/src/taker.rs [hunk 4]
```diff
@@ -168,14 +168,15 @@ async fn main() -> Result<()> {
...
+    let (wallet, wallet_fut) = wallet::Actor::new(
         opts.network.electrum(),
         &data_dir.join("taker_wallet.sqlite"),
         ext_priv_key,
     )
     .await?
     .create(None)
-    .spawn_global();
+    .run();
+    let _wallet_handle = wallet_fut.spawn_with_handle();
```
Comment (thomaseizinger) [inline daemon/src/taker.rs:179]: This is unrelated if you really only wanted to do the maker in this PR :)

File: daemon/tests/harness/mod.rs [hunk 5]
```diff
@@ -168,7 +181,8 @@ impl Taker {
         let (oracle, monitor, wallet) = mocks::create_actors(&mocks);
         mocks.mock_common_empty_handlers().await;
 
-        let wallet_addr = wallet.create(None).spawn_global();
+        let (wallet_addr, wallet_fut) = wallet.create(None).run();
+        tokio::spawn(wallet_fut);
```
Comment (thomaseizinger) [inline daemon/tests/harness/mod.rs:175]: So is this.

File: daemon/tests/harness/mod.rs [hunk 6]
```diff
@@ -153,6 +162,10 @@ impl Taker {
         &mut self.system.order_feed_receiver
     }
 
+    pub fn maker_status_feed(&mut self) -> &mut watch::Receiver<ConnectionStatus> {
+        &mut self.system.maker_online_status_feed_receiver
+    }
```
Comment (thomaseizinger) [inline daemon/tests/harness/mod.rs:157]: Is this used?

File: daemon/src/lib.rs [hunk 7]
```diff
@@ -46,16 +47,20 @@ pub mod wallet;
 pub mod wallet_sync;
 pub mod wire;
 
-const HEARTBEAT_INTERVAL: std::time::Duration = Duration::from_secs(5);
+pub const HEARTBEAT_INTERVAL: std::time::Duration = Duration::from_secs(5);
 
 pub const N_PAYOUTS: usize = 200;
 
+/// Struct controlling the lifetime of the actors
```
Comment (thomaseizinger) [inline daemon/src/lib.rs]: I think this needs more docs. Can you explain a bit here, how this works? :)

Comment (thomaseizinger) [reply daemon/src/lib.rs]: Also, I think `Tasks` might be a better name because it is not just actors but also notifier tasks.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Nice work!

A couple of changes seem to be unrelated, would you mind getting rid of those again?
Also, if you really want to make the heartbeat interval configurable, can we do that as a separate patch?

File: daemon/tests/harness/mod.rs (hunk 6 above)
Comment (thomaseizinger) [reply daemon/tests/harness/mod.rs:157]: Seems to belong to the patch after this.

File: daemon/src/lib.rs (hunk 7 above)
Comment (klochowicz) [reply daemon/src/lib.rs]: sure! `Tasks` sounds like good name. I'll describe more works.

> Nice work!
> 
> A couple of changes seem to be unrelated, would you mind getting rid of those again? Also, if you really want to make the heartbeat interval configurable, can we do that as a separate patch?



> Nice work!
> 
> A couple of changes seem to be unrelated, would you mind getting rid of those again? Also, if you really want to make the heartbeat interval configurable, can we do that as a separate patch?

Sure, I'll remove heartbeat interval changes from this one (it's not that related, it was only a test optimisation, and can go in later / if at all)

File: daemon/tests/harness/mod.rs (hunk 6 above)
Comment (klochowicz) [reply daemon/tests/harness/mod.rs:157]: moved into the correct patch.

**Comment by klochowicz** [comment]
> Nice work!
> 
> A couple of changes seem to be unrelated, would you mind getting rid of those again? Also, if you really want to make the heartbeat interval configurable, can we do that as a separate patch?

Removed the heartbeat, rolled out the same changes in the taker side. My apologies for no `fixup`s, but it was a huge change in total ;D

File: daemon/tests/harness/mod.rs [hunk 8]
```diff
@@ -64,7 +63,8 @@ impl Maker {
         let (oracle, monitor, wallet) = mocks::create_actors(&mocks);
         mocks.mock_common_empty_handlers().await;
 
-        let wallet_addr = wallet.create(None).spawn_global();
+        let (wallet_addr, wallet_fut) = wallet.create(None).run();
+        tokio::spawn(wallet_fut);
```
Comment (thomaseizinger) [inline daemon/tests/harness/mod.rs:67]: Not a blocker but it would be good to always strive for a minimal diff. This change doesn't add anything as far as I can tell?

File: daemon/tests/happy_path.rs [hunk 9]
```diff
@@ -1,9 +1,13 @@
...
+use std::time::Duration;
+
+use crate::harness::flow::{is_next_none, next, next_cfd, next_order, next_some};
 use crate::harness::{assert_is_same_order, dummy_new_order, init_tracing, start_both};
+use daemon::connection::ConnectionStatus;
 use daemon::model::cfd::CfdState;
 use daemon::model::Usd;
 use maia::secp256k1_zkp::schnorrsig;
 use rust_decimal_macros::dec;
+use tokio::time::sleep;
```
Comment (thomaseizinger) [inline daemon/tests/happy_path.rs]: I'd prefer to keep all `use` statements in a single block.

**Review by thomaseizinger — APPROVED** [review]
Nice work!

File: daemon/src/lib.rs [hunk 10]
```diff
@@ -50,12 +51,18 @@ const HEARTBEAT_INTERVAL: std::time::Duration = Duration::from_secs(5);
...
+/// If it gets dropped, all tasks are cancelled.
+pub struct Tasks(Vec<RemoteHandle<()>>);
+
 pub struct MakerActorSystem<O, M, T, W> {
     pub cfd_actor_addr: Address<maker_cfd::Actor<O, M, T, W>>,
     pub cfd_feed_receiver: watch::Receiver<Vec<Cfd>>,
     pub order_feed_receiver: watch::Receiver<Option<Order>>,
     pub update_cfd_feed_receiver: watch::Receiver<UpdateCfdProposals>,
     pub inc_conn_addr: Address<T>,
+    pub tasks_handle: Tasks,
```
Comment (thomaseizinger) [inline daemon/src/lib.rs]: Just `tasks` perhaps?

Comment (thomaseizinger) [reply daemon/src/lib.rs]: Plus make it private? Should not be modified I guess.

**Comment by thomaseizinger** [comment]
Should we ban `tokio::spawn` to avoid creating zombie tasks that are not stopped when the actor system gets dropped?

File: daemon/tests/happy_path.rs (hunk 9 above)
Comment (klochowicz) [reply daemon/tests/happy_path.rs]: I'll try to keep an eye on it (or maybe write a lint against it), this is unintentional outcome of running a code action in my emacs ;/
fixed!

File: daemon/tests/harness/mod.rs (hunk 8 above)
Comment (klochowicz) [reply daemon/tests/harness/mod.rs:67]: it's a preparation to spot the stray `tokio::spawn` calls - I will need to ban `spawn_global()` as well. 
I'll keep this change as it's minimal if you're ok with it

File: daemon/src/lib.rs (hunk 10 above)
Comment (klochowicz) [reply daemon/src/lib.rs]: renamed. 

 tried to make it private in the very beginning, but it was getting dropped if I didn't destructure it properly and keep it in scope.

An alternative would be manage the whole actor system sin Rocket, but we're not doing right now (should we?). Then it could be private, as it should be!

That's the reason I made the newtype, which at least makes the contents of `Tasks` private, so you can't add/remove/access the internals.

File: daemon/tests/harness/mod.rs (hunk 8 above)
Comment (thomaseizinger) [reply daemon/tests/harness/mod.rs:67]: We can disable the tokio feature for `xtra` then we don't have `spawn_global` available :)

**Comment by klochowicz** [comment]
> Should we ban `tokio::spawn` to avoid creating zombie tasks that are not stopped when the actor system gets dropped?

I contemplated that as well! I'll create a separate PR for that.

File: daemon/src/lib.rs (hunk 10 above)
Comment (thomaseizinger) [reply daemon/src/lib.rs]: > tried to make it private in the very beginning, but it was getting dropped if I didn't destructure it properly and keep it in scope.

Ah fair enough.

> An alternative would be manage the whole actor system sin Rocket, but we're not doing right now (should we?). Then it could be private, as it should be!

Yeah that might be worth it at some point. Could even consider making all fields private and only expose what we need.

**Review by thomaseizinger — APPROVED** [review]

File: daemon/src/lib.rs (hunk 10 above)
Comment (klochowicz) [reply daemon/src/lib.rs]: I'll put it on my list to deal with at some point.
